File: packages/merkle-tree/src/test/deleteFile.test.ts
import * as crypto from 'crypto';
import * as fs from 'fs/promises';
import { 
    addItem, 
    HashedItem, 
    IMerkleTree,
    findItemNode,
    deleteItem,
    deleteItems,
    saveTree,
    loadTree,
    createTree,
    buildMerkleTree
} from '../lib/merkle-tree';
import { FileStorage } from 'storage';

describe('File Deletion (deleteItem)', () => {

    // Helper function to create a file hash
    function createHashedItem(name: string, content: string = name): HashedItem {
        const hash = crypto.createHash('sha256')
            .update(content)
            .digest();
        return {
            name,
            hash,
            length: content.length,
            lastModified: new Date(),
        };
    }

    // Helper function to build a small test tree
    function buildTestTree(): IMerkleTree<any>{
        let tree = createTree("12345678-1234-5678-9abc-123456789abc");
        const fileNames = ['file1.txt', 'file2.txt', 'file3.txt', 'file4.txt', 'file5.txt'];
        
        for (const fileName of fileNames) {
            tree = addItem(tree, createHashedItem(fileName));
        }
        
        if (!tree) {
            throw new Error('Failed to build test tree');
        }

        tree.dirty = false;
        tree.merkle = buildMerkleTree(tree.sort); // Force tree rebuild.
        
        return tree;
    }

    test('should completely remove a file from the tree', () => {
        // Build a test tree
        const tree = buildTestTree();
        const initialNumFiles = tree.sort?.leafCount || 0;
        const initialNodeCount = tree.sort?.nodeCount || 0;
        
        // Verify the file exists before deletion
        const fileToDelete = 'file3.txt';
        const nodeBeforeDeletion = findItemNode(tree, fileToDelete);
        expect(nodeBeforeDeletion).toBeDefined();
        expect(nodeBeforeDeletion?.name).toBe(fileToDelete);
        
        // Delete the file completely
        deleteItem(tree, fileToDelete);
        
        // Verify the file is completely gone
        const nodeAfterDeletion = findItemNode(tree, fileToDelete);
        expect(nodeAfterDeletion).toBeUndefined();
        
        // Verify the tree structure has changed (fewer nodes and files)
        expect(tree.sort?.leafCount).toBe(initialNumFiles - 1);
        expect(tree.sort?.nodeCount || 0).toBeLessThan(initialNodeCount);
        
        // Verify remaining files are still present
        expect(findItemNode(tree, 'file1.txt')).toBeDefined();
        expect(findItemNode(tree, 'file2.txt')).toBeDefined();
        expect(findItemNode(tree, 'file4.txt')).toBeDefined();
        expect(findItemNode(tree, 'file5.txt')).toBeDefined();
    });

    test('should handle deleting a non-existent file', () => {
        const tree = buildTestTree();
        const initialLeafCount = tree.sort?.leafCount || 0;
        const initialNodeCount = tree.sort?.nodeCount || 0;
        
        // Attempt to delete a non-existent file
        deleteItem(tree, 'non-existent-file.txt');
        
        // Verify the tree structure is unchanged
        expect(tree.sort?.leafCount).toBe(initialLeafCount);
        expect(tree.sort?.nodeCount).toBe(initialNodeCount);
        expect(findItemNode(tree, 'non-existent-file.txt')).toBeUndefined();
    });

    test('should persist deletion when saving and loading the tree', async () => {
        // Create a test tree and delete a file
        const tree = buildTestTree();
        const fileToDelete = 'file2.txt';
        
        deleteItem(tree, fileToDelete);

        tree.dirty = false;
        tree.merkle = buildMerkleTree(tree.sort);
        
        // Save the tree to a temporary file
        const tempFile = '/tmp/merkle-tree-delete-test.bin';
        await saveTree(tempFile, tree, new FileStorage(""));
        
        // Load the tree back
        const loadedTree = (await loadTree(tempFile, new FileStorage("")))!;
        
        // Verify the file is completely gone
        expect(findItemNode(loadedTree, fileToDelete)).toBeUndefined();
        
        // Clean up
        await fs.unlink(tempFile).catch(() => {/* ignore error */});
    });

    test('should allow multiple files to be deleted', () => {
        const tree = buildTestTree();
        const initialFiles = tree.sort?.leafCount || 0;
        
        // Delete multiple files
        deleteItem(tree, 'file1.txt');
        deleteItem(tree, 'file3.txt');
        deleteItem(tree, 'file5.txt');
        
        // Check that all files are completely gone
        expect(findItemNode(tree, 'file1.txt')).toBeUndefined();
        expect(findItemNode(tree, 'file2.txt')).toBeDefined();
        expect(findItemNode(tree, 'file3.txt')).toBeUndefined();
        expect(findItemNode(tree, 'file4.txt')).toBeDefined();
        expect(findItemNode(tree, 'file5.txt')).toBeUndefined();
        
        // Verify metadata is correct
        expect(tree.sort?.leafCount).toBe(initialFiles - 3);
    });

    test('should update Merkle tree hashes when a file is deleted', () => {
        const tree = buildTestTree();
        
        // Save original root hash
        const originalRootHash = tree.merkle!.hash.toString('hex');
        
        // Delete a file
        deleteItem(tree, 'file3.txt');

        tree.dirty = false;
        tree.merkle = buildMerkleTree(tree.sort); // Force tree rebuild.
        
        // Get new root hash
        const newRootHash = tree.merkle!.hash.toString('hex');
        
        // The root hash should have changed
        expect(newRootHash).not.toBe(originalRootHash);
    });

    test('should handle deleting the only file in a tree', () => {
        // Create a tree with just one file
        let tree = createTree("12345678-1234-5678-9abc-123456789abc");
        tree = addItem(tree, createHashedItem('single-file.txt'));
        
        expect(tree.sort?.leafCount).toBe(1);
        expect(tree.sort?.nodeCount || 0).toBe(1);
        
        // Delete the only file
        deleteItem(tree, 'single-file.txt');
        
        // Tree should be empty
        expect(tree.sort?.leafCount || 0).toBe(0);
        expect(tree.sort?.nodeCount || 0).toBe(0);
    });

    test('should handle deleting from empty tree', () => {
        const emptyTree = createTree("12345678-1234-5678-9abc-123456789abc");
        
        // Attempt to delete from empty tree
        deleteItem(emptyTree, 'any-file.txt');
        
        // Tree should remain empty
        expect(emptyTree.sort?.leafCount || 0).toBe(0);
        expect(emptyTree.sort?.nodeCount || 0).toBe(0);
    });
});

describe('Hard File Deletion (deleteItems)', () => {

    // Helper function to create a file hash
    function createHashedItem(name: string, content: string = name): HashedItem {
        const hash = crypto.createHash('sha256')
            .update(content)
            .digest();
        return {
            name,
            hash,
            length: content.length,
            lastModified: new Date(),
        };
    }

    // Helper function to build a small test tree
    function buildTestTree(): IMerkleTree<any>{
        let tree = createTree("12345678-1234-5678-9abc-123456789abc");
        const fileNames = ['file1.txt', 'file2.txt', 'file3.txt', 'file4.txt', 'file5.txt'];
        
        for (const fileName of fileNames) {
            tree = addItem(tree, createHashedItem(fileName));
        }
        
        if (!tree) {
            throw new Error('Failed to build test tree');
        }

        tree.dirty = false;
        tree.merkle = buildMerkleTree(tree.sort); // Force tree rebuild.
        
        return tree;
    }

    test('should completely remove a file from the tree', () => {
        // Build a test tree
        const tree = buildTestTree();
        const initialNumFiles = tree.sort?.leafCount || 0;
        const initialNodeCount = tree.sort?.nodeCount || 0;
        
        // Verify the file exists before deletion
        const fileToDelete = 'file3.txt';
        const nodeBeforeDeletion = findItemNode(tree, fileToDelete);
        expect(nodeBeforeDeletion).toBeDefined();
        expect(nodeBeforeDeletion?.name).toBe(fileToDelete);
        
        // Delete the file completely
        const result = deleteItems(tree, [fileToDelete]);
        expect(result).toBe(1);
        
        // Verify the file is completely gone
        const nodeAfterDeletion = findItemNode(tree, fileToDelete);
        expect(nodeAfterDeletion).toBeUndefined();
        
        // Verify the tree structure has changed (fewer nodes and files)
        expect(tree.sort?.leafCount).toBe(initialNumFiles - 1);
        expect(tree.sort?.nodeCount || 0).toBeLessThan(initialNodeCount);
        
        // Verify remaining files are still present
        expect(findItemNode(tree, 'file1.txt')).toBeDefined();
        expect(findItemNode(tree, 'file2.txt')).toBeDefined();
        expect(findItemNode(tree, 'file4.txt')).toBeDefined();
        expect(findItemNode(tree, 'file5.txt')).toBeDefined();
    });

    test('should throw when trying to delete a non-existent file', () => {
        const tree = buildTestTree();
        
        expect(() => {
            deleteItems(tree, ['non-existent-file.txt']);
        }).toThrow('Cannot delete items: the following items do not exist: non-existent-file.txt');
    });

    test('should handle deleting the only file in a tree', () => {
        // Create a tree with just one file
        let tree = createTree("12345678-1234-5678-9abc-123456789abc");
        tree = addItem(tree, createHashedItem('single-file.txt'));
        
        expect(tree.sort?.leafCount).toBe(1);
        expect(tree.sort?.nodeCount || 0).toBe(1);
        
        // Delete the only file
        const result = deleteItems(tree, ['single-file.txt']);
        expect(result).toBe(1);
        
        // Tree should be empty
        expect(tree.sort?.leafCount || 0).toBe(0);
        expect(tree.sort?.nodeCount || 0).toBe(0);
    });

    test('should persist hard deletion when saving and loading the tree', async () => {
        // Create a test tree and delete a file
        const tree = buildTestTree();
        const fileToDelete = 'file2.txt';
        
        deleteItems(tree, [fileToDelete]);

        tree.dirty = false;
        tree.merkle = buildMerkleTree(tree.sort);
        
        // Save the tree to a temporary file
        const tempFile = '/tmp/merkle-tree-hard-delete-test.bin';
        await saveTree(tempFile, tree, new FileStorage(""));
        
        // Load the tree back
        const loadedTree = (await loadTree(tempFile, new FileStorage("")))!;
        
        // Verify the file is completely gone
        expect(findItemNode(loadedTree, fileToDelete)).toBeUndefined();
        
        // Clean up
        await fs.unlink(tempFile).catch(() => {/* ignore error */});
    });

    test('should allow multiple files to be deleted completely', () => {
        const tree = buildTestTree();
        const initialFiles = tree.sort?.leafCount || 0;
        
        // Delete multiple files
        const result = deleteItems(tree, ['file1.txt', 'file3.txt', 'file5.txt']);
        expect(result).toBe(3);
        
        // Check that all files are completely gone
        expect(findItemNode(tree, 'file1.txt')).toBeUndefined();
        expect(findItemNode(tree, 'file2.txt')).toBeDefined();
        expect(findItemNode(tree, 'file3.txt')).toBeUndefined();
        expect(findItemNode(tree, 'file4.txt')).toBeDefined();
        expect(findItemNode(tree, 'file5.txt')).toBeUndefined();
        
        // Verify metadata is correct
        expect(tree.sort?.leafCount).toBe(initialFiles - 3);
    });

    test('should update Merkle tree hashes when a file is deleted', () => {
        const tree = buildTestTree();
        
        // Save original root hash
        const originalRootHash = tree.merkle!.hash.toString('hex');
        
        // Delete a file
        deleteItems(tree, ['file3.txt']);

        tree.dirty = false;
        tree.merkle = buildMerkleTree(tree.sort); // Force tree rebuild.
        
        // Get new root hash
        const newRootHash = tree.merkle!.hash.toString('hex');
        
        // The root hash should have changed
        expect(newRootHash).not.toBe(originalRootHash);
    });

    test('should handle deleting all files from a tree', () => {
        const tree = buildTestTree();
        const allFiles = ['file1.txt', 'file2.txt', 'file3.txt', 'file4.txt', 'file5.txt'];
        
        // Delete all files
        const result = deleteItems(tree, allFiles);
        expect(result).toBe(allFiles.length);
        
        // Tree should be empty
        expect(tree.sort?.leafCount || 0).toBe(0);
        expect(tree.sort?.nodeCount || 0).toBe(0);
    });

    test('should preserve metadata id and update counts correctly', () => {
        const tree = buildTestTree();
        const originalMetadata = { id: tree.id, leafCount: tree.sort?.leafCount, nodeCount: tree.sort?.nodeCount, size: tree.sort?.size };
        
        // Delete a file
        deleteItems(tree, ['file3.txt']);
        
        // Check that metadata is preserved but counts are updated
        expect(tree.id).toBe(originalMetadata.id);
        expect(tree.sort?.leafCount).toBe((originalMetadata.leafCount || 0) - 1);
        expect(tree.sort?.nodeCount || 0).toBeLessThan(originalMetadata.nodeCount || 0);
        expect(tree.sort?.size || 0).toBeLessThan(originalMetadata.size || 0);
    });

    test('should throw when trying to delete from empty tree', () => {
        const emptyTree = createTree("12345678-1234-5678-9abc-123456789abc");
        
        expect(() => {
            deleteItems(emptyTree, ['any-file.txt']);
        }).toThrow('Cannot delete items from empty or invalid merkle tree');
    });

    test('should throw when trying to delete 0 files (empty array)', () => {
        const tree = buildTestTree();
        
        expect(() => {
            deleteItems(tree, []);
        }).toThrow('Cannot delete items: no names provided');
    });

    test('should handle deleting 1 file', () => {
        const tree = buildTestTree();
        const initialFiles = tree.sort?.leafCount || 0;
        
        const result = deleteItems(tree, ['file2.txt']);
        expect(result).toBe(1);
        
        // Tree should have one less file
        expect(tree.sort?.leafCount).toBe(initialFiles - 1);
        expect(findItemNode(tree, 'file2.txt')).toBeUndefined();
        expect(findItemNode(tree, 'file1.txt')).toBeDefined();
        expect(findItemNode(tree, 'file3.txt')).toBeDefined();
    });

    test('should handle deleting 2 files', () => {
        const tree = buildTestTree();
        const initialFiles = tree.sort?.leafCount || 0;
        
        const result = deleteItems(tree, ['file1.txt', 'file4.txt']);
        expect(result).toBe(2);
        
        // Tree should have two less files
        expect(tree.sort?.leafCount).toBe(initialFiles - 2);
        expect(findItemNode(tree, 'file1.txt')).toBeUndefined();
        expect(findItemNode(tree, 'file4.txt')).toBeUndefined();
        expect(findItemNode(tree, 'file2.txt')).toBeDefined();
        expect(findItemNode(tree, 'file3.txt')).toBeDefined();
        expect(findItemNode(tree, 'file5.txt')).toBeDefined();
    });

    test('should handle deleting 3 files', () => {
        const tree = buildTestTree();
        const initialFiles = tree.sort?.leafCount || 0;
        
        const result = deleteItems(tree, ['file2.txt', 'file3.txt', 'file5.txt']);
        expect(result).toBe(3);
        
        // Tree should have three less files
        expect(tree.sort?.leafCount).toBe(initialFiles - 3);
        expect(findItemNode(tree, 'file2.txt')).toBeUndefined();
        expect(findItemNode(tree, 'file3.txt')).toBeUndefined();
        expect(findItemNode(tree, 'file5.txt')).toBeUndefined();
        expect(findItemNode(tree, 'file1.txt')).toBeDefined();
        expect(findItemNode(tree, 'file4.txt')).toBeDefined();
    });

    test('should throw when trying to delete mix of existing and non-existing files', () => {
        const tree = buildTestTree();
        
        expect(() => {
            deleteItems(tree, ['file1.txt', 'non-existent.txt', 'file3.txt', 'another-missing.txt']);
        }).toThrow('Cannot delete items: the following items do not exist: non-existent.txt, another-missing.txt');
    });
});

File: packages-zig/merkle-tree-zig/src/test/deleteFile.test.zig
//
// Tests for deleteItem (port of the 'File Deletion (deleteItem)' tests of src/test/deleteFile.test.ts).
// Not ported: the 'Hard File Deletion (deleteItems)' tests (deleteItems is not ported).
//

const std = @import("std");
const merkle_tree_zig = @import("merkle-tree-zig");
const merkle_verify = @import("merkle-verify.zig");
const memory_storage = @import("memory-storage.zig");
const merkle_tree = merkle_tree_zig.merkle_tree;
const IMerkleTree = merkle_tree.IMerkleTree;

//
// Helper function to build a small test tree
//
fn buildTestTree(allocator: std.mem.Allocator) !IMerkleTree {
    var tree = merkle_tree.createTree(merkle_verify.TEST_TREE_ID);
    const fileNames = [_][]const u8{ "file1.txt", "file2.txt", "file3.txt", "file4.txt", "file5.txt" };
    for (fileNames) |fileName| {
        tree = try merkle_tree.addItem(allocator, &tree, try merkle_verify.createSha256HashedItem(allocator, fileName, fileName, fileName.len));
    }
    tree.dirty = false;
    tree.merkle = try merkle_tree.buildMerkleTree(allocator, tree.sort); // Force tree rebuild.
    return tree;
}

//
// The leaf count of a tree (0 when empty; TypeScript: `tree.sort?.leafCount || 0`).
//
fn leafCount(tree: *const IMerkleTree) u32 {
    return if (tree.sort) |sort| sort.leafCount else 0;
}

//
// The node count of a tree (0 when empty).
//
fn nodeCount(tree: *const IMerkleTree) u32 {
    return if (tree.sort) |sort| sort.nodeCount else 0;
}

test "should completely remove a file from the tree" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    // Build a test tree
    var tree = try buildTestTree(allocator);
    const initialNumFiles = leafCount(&tree);
    const initialNodeCount = nodeCount(&tree);

    // Verify the file exists before deletion
    const fileToDelete = "file3.txt";
    const nodeBeforeDeletion = merkle_tree.findItemInTree(tree.sort, fileToDelete);
    try std.testing.expectEqualStrings(fileToDelete, nodeBeforeDeletion.?.name.?);

    // Delete the file completely
    try merkle_tree.deleteItem(allocator, &tree, fileToDelete);

    // Verify the file is completely gone
    try std.testing.expect((merkle_tree.findItemInTree(tree.sort, fileToDelete)) == null);

    // Verify the tree structure has changed (fewer nodes and files)
    try std.testing.expectEqual(initialNumFiles - 1, leafCount(&tree));
    try std.testing.expect(nodeCount(&tree) < initialNodeCount);

    // Verify remaining files are still present
    for ([_][]const u8{ "file1.txt", "file2.txt", "file4.txt", "file5.txt" }) |name| {
        try std.testing.expect((merkle_tree.findItemInTree(tree.sort, name)) != null);
    }
}

test "should handle deleting a non-existent file" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = try buildTestTree(allocator);
    const initialLeafCount = leafCount(&tree);
    const initialNodeCount = nodeCount(&tree);

    // Attempt to delete a non-existent file
    try merkle_tree.deleteItem(allocator, &tree, "non-existent-file.txt");

    // Verify the tree structure is unchanged
    try std.testing.expectEqual(initialLeafCount, leafCount(&tree));
    try std.testing.expectEqual(initialNodeCount, nodeCount(&tree));
    try std.testing.expect((merkle_tree.findItemInTree(tree.sort, "non-existent-file.txt")) == null);
}

test "should persist deletion when saving and loading the tree" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    // Create a test tree and delete a file
    var tree = try buildTestTree(allocator);
    const fileToDelete = "file2.txt";
    try merkle_tree.deleteItem(allocator, &tree, fileToDelete);
    tree.dirty = false;
    tree.merkle = try merkle_tree.buildMerkleTree(allocator, tree.sort);

    // Save the tree (in memory instead of a temporary file)
    var storage = memory_storage.MemoryStorage.init(allocator);
    try merkle_tree.saveTree(allocator, std.testing.io, "merkle-tree-delete-test.bin", &tree, storage.asStorage(), "FTRE");

    // Load the tree back
    const loadedTree = (try merkle_tree.loadTree(allocator, std.testing.io, "merkle-tree-delete-test.bin", storage.asStorage(), "FTRE")).?;

    // Verify the file is completely gone
    try std.testing.expect((merkle_tree.findItemInTree(loadedTree.sort, fileToDelete)) == null);
}

test "should allow multiple files to be deleted" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = try buildTestTree(allocator);
    const initialFiles = leafCount(&tree);

    // Delete multiple files
    try merkle_tree.deleteItem(allocator, &tree, "file1.txt");
    try merkle_tree.deleteItem(allocator, &tree, "file3.txt");
    try merkle_tree.deleteItem(allocator, &tree, "file5.txt");

    // Check that all files are completely gone
    try std.testing.expect((merkle_tree.findItemInTree(tree.sort, "file1.txt")) == null);
    try std.testing.expect((merkle_tree.findItemInTree(tree.sort, "file2.txt")) != null);
    try std.testing.expect((merkle_tree.findItemInTree(tree.sort, "file3.txt")) == null);
    try std.testing.expect((merkle_tree.findItemInTree(tree.sort, "file4.txt")) != null);
    try std.testing.expect((merkle_tree.findItemInTree(tree.sort, "file5.txt")) == null);

    // Verify metadata is correct
    try std.testing.expectEqual(initialFiles - 3, leafCount(&tree));
}

test "should update Merkle tree hashes when a file is deleted" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = try buildTestTree(allocator);

    // Save original root hash
    const originalRootHash = tree.merkle.?.hash;

    // Delete a file
    try merkle_tree.deleteItem(allocator, &tree, "file3.txt");
    tree.dirty = false;
    tree.merkle = try merkle_tree.buildMerkleTree(allocator, tree.sort); // Force tree rebuild.

    // The root hash should have changed
    try std.testing.expect(!std.mem.eql(u8, originalRootHash, tree.merkle.?.hash));
}

test "should handle deleting the only file in a tree" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    // Create a tree with just one file
    var tree = merkle_tree.createTree(merkle_verify.TEST_TREE_ID);
    tree = try merkle_tree.addItem(allocator, &tree, try merkle_verify.createSha256HashedItem(allocator, "single-file.txt", "single-file.txt", 15));
    try std.testing.expectEqual(@as(u32, 1), leafCount(&tree));
    try std.testing.expectEqual(@as(u32, 1), nodeCount(&tree));

    // Delete the only file
    try merkle_tree.deleteItem(allocator, &tree, "single-file.txt");

    // Tree should be empty
    try std.testing.expectEqual(@as(u32, 0), leafCount(&tree));
    try std.testing.expectEqual(@as(u32, 0), nodeCount(&tree));
}

test "should handle deleting from empty tree" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var emptyTree = merkle_tree.createTree(merkle_verify.TEST_TREE_ID);

    // Attempt to delete from empty tree
    try merkle_tree.deleteItem(allocator, &emptyTree, "any-file.txt");

    // Tree should remain empty
    try std.testing.expectEqual(@as(u32, 0), leafCount(&emptyTree));
    try std.testing.expectEqual(@as(u32, 0), nodeCount(&emptyTree));
    try std.testing.expect(emptyTree.dirty);
}
